parsl/executors/high_throughput/interchange.py

```diff
@@ -0,0 +1,310 @@
+                        logger.debug("There are no more tasks in the incoming queue. Breaking")
+                        break
+                    # msg = self.task_incoming.recv_string()
+                    if msg == 'STOP':
+                        raise ShutdownRequest
+                    else:
+                        tasks.append(msg)
+            else:
+                logger.debug("[GET_TASKS] Returning with {} tasks".format(c))
+                break
+
+        return tasks
+
+    def start(self, poll_period=0.1):
+        """ Start the NeedNameQeueu
+
+        Parameters:
+        ----------
+
+        poll_period : float
+              Poll period in seconds
+
+        TODO: Move task receiving to a thread
+        """
+        logger.info("Incoming ports bound")
+
+        # start = time.time()
+        start = None
+        count = 0
+
+        poller = zmq.Poller()
+        poller.register(self.task_incoming, zmq.POLLIN)
+        poller.register(self.task_outgoing, zmq.POLLIN)
+        poller.register(self.results_incoming, zmq.POLLIN)
+
+        while True:
+            self.socks = dict(poller.poll(timeout=poll_period))
+
+            # Listen for requests for work
+            if self.task_outgoing in self.socks and self.socks[self.task_outgoing] == zmq.POLLIN:
+                message = self.task_outgoing.recv_multipart()
+                worker = message[0]
+                tasks_requested = int.from_bytes(message[1], "little")
+                worker = int.from_bytes(message[0], "little")
+
+                logger.debug("[MAIN] Worker[{}] requested {} tasks".format(worker, tasks_requested))
+                if worker not in self._ready_worker_queue:
+                    logger.debug("[MAIN] Adding worker to ready queue")
+                    self._ready_worker_queue[worker] = {'last': time.time(),
+                                                        'wtime': 60,
+                                                        'free_slots': tasks_requested,
+                                                        'tasks': []}
+                else:
+                    self._ready_worker_queue[worker]['last'] = time.time()
+                    self._ready_worker_queue[worker]['free_slots'] = tasks_requested
+
+            # If we had received any requests, check if there are tasks that could be passed
+            for worker in self._ready_worker_queue:
+                if self._ready_worker_queue[worker]['free_slots']:
+                    tasks = self.get_tasks(self._ready_worker_queue[worker]['free_slots'])
+                    if tasks:
+                        self.task_outgoing.send_multipart([message[0], b'', pickle.dumps(tasks)])
+                        tids = [t['task_id'] for t in tasks]
+                        logger.debug("[MAIN] Sent tasks: {} to {}".format(tids, worker))
+                        self._ready_worker_queue[worker]['free_slots'] -= len(tasks)
```

Comment:
The worker has a thread that sends a message with the count of waiting workers continuously. As soon as tasks are received the number of available workers goes down. And the advertised capacity is reduced. There's a possibility for a race condition since the workers don't wait for a response to each request for tasks and accidentally could be sent more task batches.